devicetree: document and test includes at child-binding level

gmarull opened the pull request:
It looks like it was possible to include other bindings at any level other than root. However, this feature was not tested nor documented.

Example:
```yaml
   # foo.yaml will be merged with content at this level
   include: foo.yaml
   child-binding:
     # bar.yaml will be merged with content at this level
     include: bar.yaml
```

gmarull committed:
devicetree: edtlib: tests: add coverage for child includes

Add test coverage for the child-binding include feature. It includes
verification of included properties as well as usage of allow/blocklist.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic commented on dts/bindings/pinctrl/atmel,sam0-pinctrl.yaml:
```diff
@@ -105,6 +95,17 @@ child-binding:
+        property-allowlist:
+          - bias-pull-up
+          - bias-pull-down
+          - drive-open-drain
```
This one isn't in the allowlist that's being removed above, and therefore should not be here, right?

gmarull replied:
Oops my mistake

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node-group.yaml:
```diff
@@ -1,165 +0,0 @@
-description: |
-  Generic pin configuration schema.
-
-  This file is equivalent to pincfg-node.yaml but inserts properties at a
-  grandchildren level, useful for group based representation.
```
Don't you think we should keep this here, for out of tree users who are depending on it?

Not blocking on this one since I no longer maintain bindings, but I would strongly suggest it.

mbolivar-nordic replied:
> I’d be surprised if there are pinctrl oot users, since it’s a de-facto requirement for any platform.

There are out of tree socs and even out-of-tree architectures. These are first class citizens in Zephyr. I understand you want to make changes quickly but I would not agree that the new treewide change policy is "unnecessary" and neither did the rest of the TSC.

mbolivar-nordic commented on dts/bindings/pinctrl/pincfg-node.yaml:
```diff
@@ -1,160 +1,159 @@
+    https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml
+
+properties:
+  bias-disable:
```
Again, you are breaking out of tree users by promoting these to top level.

mbolivar-nordic reviewed (changes requested):
Clever idea about the bindings language extension in the python-devicetree package. I am +1 for the idea, but please update the documentation in doc/build/dts/bindings.rst as well before we can merge this.

I found what looks like one mistake in the bindings changes.

Since the current changes to the bindings are breaking for out of tree users, I think this is pretty clearly a treewide change, so I'm going to apply that label. If you create a new bindings include file that's meant to be pulled in from the child-binding level instead, that would be backwards compatible and I'd be fine with removing the label in that case.

gmarull commented:
> Clever idea about the bindings language extension in the python-devicetree package. I am +1 for the idea, but please update the documentation in doc/build/dts/bindings.rst as well before we can merge this.
> 

Updated docs

> I found what looks like one mistake in the bindings changes.
>

Fixed
 
> Since the current changes to the bindings are breaking for out of tree users, I think this is pretty clearly a treewide change, so I'm going to apply that label. If you create a new bindings include file that's meant to be pulled in from the child-binding level instead, that would be backwards compatible and I'd be fine with removing the label in that case.

As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.

mbolivar-nordic commented:
> As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.

Fine, but this is still a treewide change and subject to the associated requirements.

mbolivar-nordic reviewed (dismissed):
+1 from me on the DT infrastructure. Will add DNM just to be extra careful since it's treewide and hasn't gone through arch WG yet

gmarull commented:
> > As mentioned, -1 on going that path, this is a localized change to pinctrl, an unstable API.
> 
> Fine, but this is still a treewide change and subject to the associated requirements.

why is something affecting a single driver class treewide?

mbolivar-nordic commented:
> why is something affecting a single driver class treewide?

I think this is addressed by the definition of treewide here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

carlescufi commented:
> > why is something affecting a single driver class treewide?
> 
> I think this is addressed by the definition of treewide here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

I have to agree with @gmarull here. This is just too narrow in scope to fit that definition in my opinion.

galak reviewed (changes requested):
Can you add more tests to verify that `property-allowlist` and `property-blocklist` work there.

gmarull commented:
> Can you add more tests to verify that `property-allowlist` and `property-blocklist` work there.

done

gmarull commented:
Added TSC label to discuss wether this applies for treewide or not.

gmarull committed:
doc: build: dts: mention it is allowed to include at any level

This feature was not mentioned, even though it is supported.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
I realized we actually support this feature, but it was not documented or tested. Updated this PR with tests/docs only, and splitted pinctrl part here https://github.com/zephyrproject-rtos/zephyr/pull/52290. This is no longer a treewide change.
